light-client/src/client_state.rs

```diff
@@ -0,0 +1,400 @@
+use prost::Message;
+
+pub const OPTIMISM_CLIENT_STATE_TYPE_URL: &str = "/ibc.lightclients.optimism.v1.ClientState";
+
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct ClientState {
+    pub chain_id: u64,
+
+    /// IBC Solidity parameters
+    pub ibc_store_address: Address,
+    pub ibc_commitments_slot: H256,
+
+    ///Light Client parameters
+    pub trusting_period: Duration,
+    pub max_clock_drift: Duration,
+
+    /// State
+    pub latest_height: Height,
+    pub frozen: bool,
+
+    /// L2 RollupConfig
+    pub rollup_config: RollupConfig,
+
+    /// L1 Config
+    pub l1_config: L1Config,
+}
+
+impl ClientState {
+    /// canonicalize canonicalizes some fields of specified client state
+    /// target fields: latest_height, frozen
+    pub fn canonicalize(mut self) -> Self {
+        self.latest_height = Height::new(0, 0);
+        self.frozen = false;
+        self
+    }
+
+    pub fn freeze(mut self) -> Self {
+        self.frozen = true;
+        self
+    }
+
+    pub fn check_header_and_update_state<const L1_SYNC_COMMITTEE_SIZE: usize>(
+        &self,
+        now: Time,
+        trusted_consensus_state: &ConsensusState,
+        header: Header<L1_SYNC_COMMITTEE_SIZE>,
+    ) -> Result<(ClientState, ConsensusState, Height, Time), Error> {
+        // Since the L1 block hash is used for L2 derivation, the validity of L1 must be verified.
+        let l1_consensus = header.verify_l1(
+            &self.l1_config,
+            now.as_unix_timestamp_secs(),
+            trusted_consensus_state,
+        )?;
+
+        // Ensure L2 header is valid
+        let (l2_header, l2_output_root) = header.verify_l2(
+            self.chain_id,
+            trusted_consensus_state.output_root,
+            &self.rollup_config,
+        )?;
+
+        // Ensure account storage is valid
+        header.account_update.verify_account_storage(
+            &self.ibc_store_address,
+            H256::from_slice(l2_header.state_root.0.as_slice()),
+        )?;
+
+        // check if the current timestamp is within the trusting period
+        validate_state_timestamp_within_trusting_period(
+            now,
+            self.trusting_period,
+            trusted_consensus_state.timestamp,
+        )?;
+        // check if the header timestamp does not indicate a future time
+        let timestamp = new_timestamp(l2_header.timestamp)?;
+        validate_header_timestamp_not_future(now, self.max_clock_drift, timestamp)?;
+
+        let mut new_client_state = self.clone();
+        let header_height = Height::new(header.trusted_height.revision_number(), l2_header.number);
+        if new_client_state.latest_height < header_height {
+            new_client_state.latest_height = header_height;
+        }
+        let new_consensus_state = ConsensusState {
+            storage_root: header.account_update.account_storage_root,
+            timestamp: new_timestamp(l2_header.timestamp)?,
```

Comment:
The current implementation verifies trusting period based on L2 timestamp, but I think that the concept of "trusting period" comes from the staking period of PoS. Therefore we should verify trusting period based on L1 timestamp, right?